### PR TITLE
Project Structure Update

### PR DESCRIPTION
**Pull request by kleinreact**
This PR re-structures the project as follows:
* `bittide` _(library)_: hardware descriptions (target independent, loosely connected / standalone)
* `bittide-extra` _(library)_:  everything general purpose that we are missing from existing upstream libraries
* `bittide-experiments` _(library)_: all infra / configs for experiments that we like to run
* `bittide-instances` _(library)_:  top-entities for synthesis (fixing target specific parameters and connecting independent components to architectures)
* `bittide-shake` _(library)_: build and CI infrastructure
* `bittide-tools` _(executables)_: all tools required for development / CI / testing / and so forth

The clean separation between libraries and executables helps to find things quickly and to avoid redundancies (as it happened with the `clash` binaries before). It cannot cause problems to have all executables at the end of the dependency chain, as libraries never can depend on executables. The dependencies of each executable still can be specified for each entry individually in `bittide-tools`.

Furthermore, the following minor has been optimized:
* `elastic-buffer-sim` has been migrated to be part of `bittide-experiments`.
* `get-data-file-name` has not necessary any more. Paths can be accessed directly via the respective libraries instead.
* All redundant `clash` executables have been removed. `bittide-tools` offers the only instance and the desired dependencies can be specified there.
* Some minor bugs with the `doctest` instances have been fixed on the way.

### CODE REVIEW

**Review by martijnbastiaan — COMMENTED** [review]
I agree with and like most of the changes here. The introduction of `bittide-experiments` (and the migration of `elastic-buffer-sim`) looks like a good idea. The redundant `clash` binaries of course had to go too. Could you check whether the descriptions (in the Cabal file) of each package conforms to what you've posted in this PR's cover letter? I think the descriptions you put here are spot-on.

Unfortunately, I do also disagree with parts of this PR. I hadn't understood from our discussion on Slack that `bittide-tools` would be the only package hosting executables. I don't think this is the way to go. A package in the Haskell world is widely understood to host both libraries and executables, just for that reason I don't think we should deviate from that. However, I've also got some more fundamental points:

* We should strive\* for packages that exist on their own, i.e., would be publishable to Hackage in their own right. That means `bittide-shake` should have the executable, because that's the whole point of the package. Similarly, `program-stream` is perfectly usable on its own. Given that we don't have a separate package for it though, `bittide-tools` is the right place. Again similarly, I think `clash` should be hosted in `bittide-instances`, but for much more pragmatic reasons: introducing another package that hosts instances would become a pain. Either the `clash` in `bittide-tools` would have to depend on both `bittide-instances` and the newly introduced package, or you'd have to prefix them. Neither of which solutions are preferential above just hosting them in their respective packages.
* More fundamentally; I think "stuff" should be grouped based on concepts, not types. For example, type classes in a Haskell project should be hosted next to where they're relevant. A class `Foo` related `XException`s should be hosted in `Clash.XException` (or `Clash.XException.Foo`), not `Clash.Classes.Foo`. A slightly more rediculous example: we don't group all `data`, `type`, and function definitions separately - we group them based on their mutual relevance! Similarly, libraries and executables should be organized such that they're conceptually close, not _because_ they're libraries and executables. This makes it much more easy to find relevant resources.

The last point also ties into were I disagree on the reasoning _for_ splitting it on libraries and executables:

> The clean separation between libraries and executables helps to find things quickly and to avoid redundancies (as it happened with the clash binaries before).

I don't think this was hard.. you could just do `rg ^executable`. On the flip side, finding out which executables are relevant for which package is now much harder! For the Clash binaries: yeah, this was a copy+pasta error basically. Clash starter projects have these executables, and when introducing the new package we simply forgot to remove them. This can happen whether we have `bittide-tools` not, we just need competent reviewers ;-).

In short: I think `bittide-tools` should exist, but only for tools/executables that don't have a clear home otherwise. Otherwise it'll just become a dumping ground / "God" package.

(Also it wasn't my intention to write half a book on this.. :smile_cat: )

File: bittide-experiments/bittide-experiments.cabal [hunk 1]
```diff
@@ -1,11 +1,11 @@
 cabal-version:       2.4
-name:                elastic-buffer-sim
+name:                bittide-experiments
 version:             0.1
 License:             Apache-2.0
 license-file:        LICENSE
 author:              QBayLogic B.V.
 maintainer:          devops@qbaylogic.com
-Copyright:           Copyright © 2022 Google LLC
+Copyright:           Copyright © 2024 Google LLC
```
[the remark this answers is not in the document]
Comment (kleinreact) [reply bittide-experiments/bittide-experiments.cabal:8]: There was no `bittide-experiments.cabal` before 2024. I probably just forgot to update it after a copy paste.

[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply bittide-experiments/bittide-experiments.cabal:8]: Ah gotcha! I don't understand why it shows up like this then.

[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply bittide-experiments/bittide-experiments.cabal:8]: Oh because it _moved_ from elastic-buffer-sim?

**Comment by martijnbastiaan** [comment]
> We should strive* for packages that exist on their own

Forgot to write the foot mark: I understand that this is hard and will probably not happen within Bittide any time soon. But I do think the principles should still apply.

File: bittide-experiments/bittide-experiments.cabal (hunk 1 above)
[the remark this answers is not in the document]
Comment (kleinreact) [reply bittide-experiments/bittide-experiments.cabal:8]: Probably. So I guess it can be handled both, either as a new file with fresh copyright or an adapted old one. Whatever you like it to be.

[the remark this answers is not in the document]
Comment (martijnbastiaan) [reply bittide-experiments/bittide-experiments.cabal:8]: I don't mind either way, we can leave it as is.

**Comment by kleinreact** [comment]
Thanks for your insights, but although it may look like that we disagree, I think we both are on the same track here. It's just that we have to align a bit our conceptional view on "what is a package" and what "grouping things together" actually means. So let me elaborate a bit on my perspective and things hopefully become clearer after that.

To dive straight in, a package should be a group of conceptionally similar functionality that can be shared in form of a library or an executable for reuse by others. Whether it is a library or an executable only depends on the desired modus operandi, so whether functionality is required at the code level (other libraries / executables), or at the system level (terminal / scripts / CI). This is also why `cabal` allows to mix up libraries and executables, as the same functionality could be offered via an API and a CLI in tandem (a library in the former and an executable in the latter case). Both will clearly end up in the same package in the end.

But that is only one side of the coin. The other side considers the management principle that everything shipped with a package should serve the usage of the package on its own. More particular, a package should not contain artifacts that are of no use without further packages coming later in the dependency chain, e.g., if a package `A` ships program `x`, but the only way to really use `x` in some context also requires package `B` (where `B` depends on `A`, but not `A` on `B`), then `x`  should either be part of package `B` or some other package `C` depending on `A` and `B`. This principle aligns with the idea that conceptual things are grouped together, but at the same time ensures that groups are chosen in such a way that they only introduce backward dependencies, but no forward ones. We don't like to introduce circular dependencies, I think that is for sure.

Now, let's make this concrete to the case of the Bittide project. Here, `bittide`, `bittide-extra`, `bittide-experiments` and `bittide-instances` clearly come with the character of a library, as their primary purpose is to be used as such. Note that I especially use the wording "character of" here to differentiate from the management artifacts required by cabal for packaging, which can be different than the character of the package itself.

Lets have a look at the `hitl` executable living in `bittide-instances` in the original setup then. This executable mainly uses functionality of `bittide-instances`, so it may look natural that it should be bundled with that package. However, only having `bittide-instances` and `hitl` at hand is of no use, as the functionality of `hitl` is tightly connected to the HITL test environment, which also requires `bittide-shake` (`HardwareTest.tcl` in particular) and basically the hole Bittide project to make sense. Therefore, only `bittide-tools` or a distinct package with executable character are valid options according to the aforementioned principles. Looking at the current complexity and size of `hitl` turns `bittide-tools` into the final candidate, as a distinct package would be overkill right now.

Lets move on to `bittide-shake`, which is an interesting candidate, as it depends a bit on the way you like to look at it. If we consider the package to have the character of a library, then the `shake` executable should live in `bittide-tools`, as again there is not use of the `shake` binary alone without the other artifacts of the whole Bittide project in the first place. On the other hand, we could say that `bittide-shake` has the character of an executable as well, making it a direct dependency of the Bittide project instead. It's just an advanced enough executable to deserve it's on package. 

I considered `bittide-shake` to have library character at first, but you are right, it's probably better to consider it as a package with executable character, as the cabal-library part of it could also live in the cabal executable section instead. We only split it apart as cabal requires this for linking the test suites to the functions to be tested. 

Finally, note that the Bittide project is also a package on its own (not a cabal one tough) depending on `bittide`, `bittide-extra`, `bittide-experiments` and `bittide-instances` (with respect to their library character) and shipping the executables `shake` and everything in `bittide-tools`, making it the last entry in the dependency chain. This especially makes `bittide-tools` **not** the sink for all the executables, but more a container for all executables that only make sense in the scope of the Bittide project as a whole.

I hope that clarifies my thoughts and applied principles a bit. Nevertheless, it still may be that I have missed some fact, looked at things the wrong way, or just described things poorly during the discussion so far. So please be critical and ask if decisions or changes are still unclear to you.

**Comment by martijnbastiaan** [comment]
Okay, clear, good to see we agree. So let's:

* Move `shake` back to `bittide-shake`. As said the sole purpose of `bittide-shake` is to provide this executable.
* Move `clash` back to `bittide-instances`. This leaves it open to introduce Clash top entities in other packages (for reasons mentioned earlier). Also, it keeps the structure of the starter project, which is valuable in its own right IMO. (Note that it's a little bit iffy we have to duplicate `clash` this way, but the alternative would be writing our own Clash specific Cabal, which is not something we'll do anytime soon / ever.)
* Given that `post-board-test-extended` is really part of the test (-instance) in `bittide-instances`, I think we should move it back there too. Otherwise we scatter (test specific) test code across packages.

For the other ones I'm swayed by your arguments.

**Review by martijnbastiaan — CHANGES_REQUESTED** [review]
So I'll put in a formal "request changes", but LGTM overall :-).

**Review by martijnbastiaan — APPROVED** [review]
